tour_model_eval/confirmed_trips_eval_bins_clusters.py

```diff
@@ -20,54 +22,213 @@
+    valid_trips = [non_empty_trips[i]for i in valid_trips_idx_ls]
+
+    sim = similarity.similarity(valid_trips, radius)
+    filter_trips = sim.data
+    return filter_trips,sim,trips
+
+def valid_user(filter_trips,trips):
+    valid = False
+    if len(filter_trips) >= 10 and len(filter_trips) / len(trips) >= 0.5:
+        valid = True
+    return valid
+
+
+# v_measure_bins takes 5 parameters
+# - sp2en=True: change Spanish to English
+# - cvt_pur_mo=True: convert purposes and replaced mode
+# - cutoff=True: choose to analyze bins above cutoff
+# - cutoff=None: analyze all bins
+# Note: for sp2en and cvt_pur_mo, set either one to be True as needed. cvt_pur_mo will change language first
+def v_measure_bins(all_users,radius,sp2en=None,cvt_pur_mo=None,cutoff=None):
+    homo_score = []
+    comp_score = []
+    v_score = []
     for i in range(len(all_users)):
         user = all_users[i]
-        trips = pipeline.read_data(uuid=user, key=esda.CONFIRMED_TRIP_KEY)
-        all_bins_preci = []
-        non_empty_trips = [t for t in trips if t["data"]["user_input"] != {}]
-        if non_empty_trips != {}:
-            sim = similarity.similarity(non_empty_trips, radius)
-            if sim.data:
-                sim.bin_data()
-                all_bins_preci = precision_bins(all_bins_preci, sim.bins, non_empty_trips, sp2en, cvt_purpose)
-        all_users_preci.append(round(mean(all_bins_preci), 2))
-    return all_users_preci
+        filter_trips,sim,trips = filter_data(user,radius)
+
+        # filter out users that haven't enough trips (at least 10 valid trips
+        # and 50% of total trips are valid) to analyze
+        if not valid_user(filter_trips,trips):
+            homo_score.append(NaN)
+            comp_score.append(NaN)
+            v_score.append(NaN)
+            continue
+        sim.bin_data()
+        if cutoff is None:
+            trip_index_ls = []
+            bins = sim.bins
+            for bin in bins:
+                for index in bin:
+                    trip_index_ls.append(index)
+            bin_trips = [filter_trips[num] for num in trip_index_ls]
+
+        elif cutoff:
+            sim.delete_bins()
+            bin_trips = sim.newdata
+            bins = sim.bins
+
+        bin_trips_df = pd.DataFrame(data=[trip["data"]["user_input"] for trip in bin_trips])
+
+        if sp2en:
+            bin_trips_df = bin_trips_df.replace(span_eng_dict)
+        elif cvt_pur_mo:
+            bin_trips_df = bin_trips_df.replace(span_eng_dict)
+            bin_trips_df = bin_trips_df.replace(map_pur_dict)
+            for a in range(len(bin_trips_df)):
+                if bin_trips_df.iloc[a]["replaced_mode"] == "same_mode":
+                    # to see which row will be converted
+                    logging.debug("The following rows will be changed: %s", bin_trips_df.iloc[a])
+                    bin_trips_df.iloc[a]["replaced_mode"] = bin_trips_df.iloc[a]['mode_confirm']
```

Comment:
This code has been duplicated twice - seems like it should be possible to pull it out.
Ideally, you would set this up as a pipeline in which this filtering happens first before the binning/clustering